node/src/store.rs

```diff
@@ -1,9 +1,136 @@
+use celestia_types::ExtendedHeader;
+use dashmap::mapref::entry::Entry;
+use dashmap::DashMap;
+use tendermint::Hash;
+use thiserror::Error;
+use tracing::{error, info};
+
+use crate::exchange::ExtendedHeaderExt;
+
 #[derive(Debug)]
-pub struct Store {}
+pub struct Store {
+    headers: DashMap<Hash, ExtendedHeader>,
+    height_to_hash: DashMap<u64, Hash>,
+}
+
+#[derive(Error, Debug, PartialEq)]
+pub enum InsertError {
+    // TODO: do we care about the distinction
+    #[error("Hash {0} already exists in store")]
+    HashExists(Hash),
+    #[error("Height {0} already exists in store")]
+    HeightExists(u64),
+}
+
+#[derive(Error, Debug, PartialEq)]
+pub enum ReadError {
+    // TODO: should we roll internal errors into one
+    #[error("Store in inconsistent state, lost head")]
+    LostStoreHead,
+    #[error("Store in inconsistent state, height->hash mapping exists, but hash doesn't")]
+    LostHash,
+    #[error("Failed to convert height from usize to u64, should not happen")]
+    HeadHeightConversionError,
+
+    #[error("Header not found in store")]
+    NotFound,
+}
 
 impl Store {
     pub fn new() -> Self {
-        Store {}
+        Store {
+            headers: DashMap::new(),
+            height_to_hash: DashMap::new(),
+        }
+    }
+
+    pub fn get_head_height(&self) -> usize {
+        self.height_to_hash.len()
+    }
+
+    pub fn append_header(&self, header: ExtendedHeader) -> Result<(), InsertError> {
+        let hash = header.hash();
+        let height = header.height(); // TODO: should Store be responsible for making sure we have
+                                      // all 1..=head headers?
+
+        // lock both maps to ensure consistency
+        // this shouldn't deadlock as long as we don't hold references across awaits if any
+        // https://github.com/xacrimon/dashmap/issues/233
+        let hash_entry = self.headers.entry(hash);
+        let height_entry = self.height_to_hash.entry(height.into());
+
+        if matches!(hash_entry, Entry::Occupied(_)) {
+            return Err(InsertError::HashExists(hash));
+        }
+        if matches!(height_entry, Entry::Occupied(_)) {
+            return Err(InsertError::HeightExists(height.into()));
+        }
+
+        info!("Will insert {hash} at {height}");
+        hash_entry.insert(header);
+        height_entry.insert(hash);
+
+        Ok(())
+    }
+
+    pub fn get_head(&self) -> Result<ExtendedHeader, ReadError> {
+        let head_height = match self.height_to_hash.len() {
```

Comment:
Based on my comment about `1..=head`, this can not be done. The highest know height number must be save in a variable and used directly. If you are going the use the continues range idea, it will be the `end` of the latest range.